src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -1087,6 +1108,127 @@ private static FunctionResult LoadJsonContentResultBuilder(IBinder binder, IFile
+            }
+
+            if (tokenSelectorPath is not null)
+            {
+                try
+                {
+                    var selectTokens = token.SelectTokens(tokenSelectorPath, false).ToList();
+                    switch (selectTokens.Count)
+                    {
+                        case 0:
+                            return new(ErrorType.Create(DiagnosticBuilder.ForPosition(arguments[1]).NoJsonTokenOnPathOrPathInvalid()));
+                        case 1:
+                            token = selectTokens.First();
+                            break;
+                        default:
+                            token = new JArray();
+                            foreach (var selectToken in selectTokens)
+                            {
+                                ((JArray)token).Add(selectToken);
+                            }
+                            break;
+                    }
+                }
+                catch (JsonException)
+                {
+                    //path is invalid or user hasn't finished typing it yet
+                    return new(ErrorType.Create(DiagnosticBuilder.ForPosition(arguments[1]).NoJsonTokenOnPathOrPathInvalid()));
+                }
+            }
+
+            return new(ConvertJsonToBicepType(token), ConvertJsonToExpression(token));
+        }
+
+        [Obsolete("This method has been replaced by ExtractTokenFromObject which supports both YAML and JSON")]
+        public static JToken OldExtractTokenFromObject(string fileContent)
+        {
+            return fileContent.TryFromJson<JToken>();
+        }
+
+        private static void CastPrimiteTypes(JToken jtoken)
+        {
+            switch (jtoken)
+            {
+                case JArray jArray:
+                    CastArray(jArray);
+                    break;
+                case JObject jObject:
+                    CastObject(jObject);
+                    break;
+                default:
+                    {
+                        var token = jtoken as JValue;
+                        if (bool.TryParse((string?)token, out bool boolean))
+                        {
+                            token.Replace(boolean);
+                        }
+                        else if (long.TryParse((string?)token, out long num))
+                        {
+                            token.Replace(num);
+                        }
+                        break;
+                    }
+            }
+        }
+
+        private static void CastArray(JArray jArray)
+        {
+            for (int i = 0; i < jArray.Count; i++)
+            {
+                CastPrimiteTypes(jArray[i]);
+            }
+
+        }
+
+        private static void CastObject(JObject jObject)
+        {
+            foreach (var item in jObject)
+            {
+                CastPrimiteTypes(item.Value!);
+            }
+
+        }
+
+        public static JToken ExtractTokenFromObject(string fileContent)
```

Comment:
Please could you move all the YAML parsing logic into a dedicated helper class? (`YamlHelper.ConvertToJToken(...)`)?

Reply:
we may have gone to far with this... we also extracted the overlapping JSON logic.

https://github.com/Azure/bicep/blob/0636d506c02381fb22a92159892c8458429c176f/src/Bicep.Core/Semantics/IObjectParser.cs#L11
https://github.com/Azure/bicep/blob/0636d506c02381fb22a92159892c8458429c176f/src/Bicep.Core/Semantics/ObjectParser.cs:L16

https://github.com/Azure/bicep/blob/0636d506c02381fb22a92159892c8458429c176f/src/Bicep.Core/Semantics/YamlObjectParser.cs#L13
https://github.com/Azure/bicep/blob/0636d506c02381fb22a92159892c8458429c176f/src/Bicep.Core/Semantics/JsonObjectParser.cs:L11